### PR TITLE
Add enum plot panel

### LINKED ISSUE #623: New Twix Panel to plot enum over time
It would be great to have a Twix panel that can show  an enum over time, so you can see state-changes nice.

So it would be a 1 dimensional Plot thingy ...
A simple bar with separators and text-labels inside would do ...

Eample:

Gamestate
playing --------------| ready -----| set --| playing ----------------------|

### PR DESCRIPTION
**Pull request by Narcha**
## Introduced Changes

This PR adds a twix panel that visualizes state changes.

Fixes #623

## Ideas for Next Iterations (Not This PR)

Improve the plot segments' functionality using [APIs](https://github.com/emilk/egui/pull/4193) recently made public

## How to Test

Twix.
Plot for example `Control.main_outputs.primary_state` or `Control.main_outputs.fall_state`

### CODE REVIEW

File: tools/twix/src/panels/enum_plot.rs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+            viewport_width: 1.0,
+            output_key: String::new(),
+            viewport_mode: ViewportMode::Full,
+        }
+    }
+}
+
+impl EnumPlotPanel {
+    fn plot_segment(plot_ui: &mut PlotUi, segment: &Segment) {
+        // Leave small gaps on the top and bottom,
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:120]: unnecessary comment?

File: tools/twix/src/panels/enum_plot.rs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        // to avoid clipping half the segment border
+        const VERTICAL_MARGIN: f64 = 0.05;
+        const BORDER_WIDTH: f32 = 2.0;
+
+        let name = segment.name();
+        let color = color_hash(&name);
+
+        let start = segment.start as f64;
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:128]: just save as f64 directly?

File: tools/twix/src/panels/enum_plot.rs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+            .allow_zoom(false)
+            .label_formatter(|_name, plot_point| format!("Message #{}", plot_point.x as isize))
+            .show(ui, |plot_ui| self.show_plot(plot_ui));
+
+        if plot.response.double_clicked() {
+            self.viewport_width = (self.messages_count as f64).max(1.0);
+            self.scroll_pos = 0.0;
+        }
+
+        if let Some(hover_pos) = plot.response.hover_pos() {
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:263]: pda ;)

File: tools/twix/src/panels/enum_plot.rs [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                .iter()
+                .find(|segment| {
+                    (segment.start as f64) < plot_hover_pos && plot_hover_pos < (segment.end as f64)
+                })
+                .cloned();
+
+            if let Some(tooltip) = hovered_segment.and_then(|segment| segment.tooltip()) {
+                show_tooltip_at_pointer(
+                    ui.ctx(),
+                    eframe::egui::Id::new("enum_plot_segment_tooltip"),
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:277]: replace by use

File: tools/twix/src/panels/enum_plot.rs [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+impl Widget for &mut EnumPlotPanel {
+    fn ui(self, ui: &mut Ui) -> Response {
+        let error = self.update_changes().err();
+
+        ui.vertical(|ui| {
+            self.plot(ui);
+
+            if let Some(error) = error {
+                ui.label(RichText::new(error).color(Color32::RED));
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:310]: you can use ui.colored_label here

File: tools/twix/src/panels/enum_plot.rs [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            }
+
+            ui.horizontal_top(|ui| {
+                let subscription_field = ui.add(CompletionEdit::outputs(
+                    &mut self.output_key,
+                    self.nao.as_ref(),
+                ));
+
+                if subscription_field.changed() {
+                    info!("Subscribing: {}", self.output_key);
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:320]: is this a useful log message?

File: tools/twix/src/panels/enum_plot.rs [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+#[derive(Debug, PartialEq)]
+enum ViewportMode {
+    // Always keep every segment in view
+    Full,
+    // Variable-sized window, fixed on the right side to the most recent data
+    Follow,
+    // The viewport does not follow new data and has to be positioned manually
+    Free,
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:87]: we usually have a no comments policy, not sure what to do with this

File: tools/twix/src/change_buffer.rs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+#[derive(Debug)]
+pub struct ChangeBufferUpdate {
+    pub updates: Vec<Change>,
+    pub message_count: usize,
+}
+
+pub enum Message {
+    GetAndReset {
+        response_sender: oneshot::Sender<Result<ChangeBufferUpdate, String>>,
```
Comment (oleflb) [inline tools/twix/src/change_buffer.rs:28]: replace use by import

File: tools/twix/src/change_buffer.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    },
+    Reset,
+}
+
+pub struct ChangeBuffer {
+    command_sender: mpsc::Sender<Message>,
+}
+
+async fn change_buffer(
+    mut subscriber_receiver: mpsc::Receiver<SubscriberMessage>,
```
Comment (oleflb) [inline tools/twix/src/change_buffer.rs:38]: replace use by import

File: tools/twix/src/panels/enum_plot.rs [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+            Value::Null => "<Null>".into(),
+            Value::Bool(value) => value.to_string(),
+            Value::Number(_) => "<Number>".into(),
+            Value::String(string) => string.clone(),
+            Value::Array(_) => "<Array>".into(),
+            Value::Object(map) => {
+                // There is only one key in the map.
+                // This is probably an enum variant.
+                if let [(key, _value)] = &map.iter().collect::<Vec<_>>().as_slice() {
+                    (*key).clone()
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:55]: still not a huge fan of this for a few nano seconds of performance. But I guess it's fine

File: tools/twix/src/panels/enum_plot.rs [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+impl Segment {
+    fn name(&self) -> String {
+        match &self.value {
+            Value::Null => "<Null>".into(),
+            Value::Bool(value) => value.to_string(),
+            Value::Number(_) => "<Number>".into(),
+            Value::String(string) => string.clone(),
+            Value::Array(_) => "<Array>".into(),
+            Value::Object(map) => {
+                // There is only one key in the map.
```
Comment (oleflb) [inline tools/twix/src/panels/enum_plot.rs:52]: I understand the need for a comment here, but maybe move in the if block. At least how it is written suggests so

**Review by oleflb — CHANGES_REQUESTED** [review]
Nice feature, left some remarks

File: tools/twix/src/change_buffer.rs (hunk 8 above)
Comment (Narcha) [reply tools/twix/src/change_buffer.rs:28]: I disagree. Just a few lines below this (l. 34), we use a `mpsc::Sender`. Having the channel type as a prefix here is the best way to distinguish between them IMO.

File: tools/twix/src/panels/enum_plot.rs (hunk 10 above)
Comment (Narcha) [reply tools/twix/src/panels/enum_plot.rs:55]: Neither am I tbh. This was the cleanest way I found to check if there is only one key, and then to safely destructure into that key. I'm open to suggestions here :)

File: tools/twix/src/panels/enum_plot.rs (hunk 6 above)
Comment (Narcha) [reply tools/twix/src/panels/enum_plot.rs:320]: This is consistent with the plot panel
https://github.com/HULKs/hulk/blob/cef35f22440b860de66d90ee8e74f407441d23f3/tools/twix/src/panels/plot.rs#L119-L122

File: tools/twix/src/change_buffer.rs (hunk 9 above)
Comment (Narcha) [reply tools/twix/src/change_buffer.rs:38]: Same as above

**Review by oleflb — APPROVED** [review]
Nize

**Pull request merged**